예외 발생시 일관된 예외 형태가 출력될 수 있도록 초기 구성

sgo8308 opened the pull request:
### **개요**
ControllerAdvice를 이용하여 컨트롤러의 비즈니스 로직과 예외처리 관련 코드를 분리하였습니다.
미리 지정해준 패키지에 속해 있는 컨트롤러에서 발생하는 예외들 외에 모든 예외는 GlobalControllerAdvice에서 처리하도록 하였습니다.

### **이슈 링크**

#5

### **작업 사항 (변경 사항)**

* GlobalControllerAdvice, SampleControllerAdvice, SampleErrorController, SampleErrorResult 추가


### 테스트 내역 (o/x)

* user 패키지에 존재하는 SampleErrorController에서 예외 발생 시 SampleControllerAdvice에서 정의한@ExceptionHandler가 붙어 있는 메소드가 받아서 에러 메시지를 구성하고 상태코드와 함께 정상적으로 json 형태 메세지를 리턴하는 것을 확인하였습니다.

### **리뷰 받고 싶은 내용**

sgo8308 committed:
feat(user, common): 예외 발생시 일관된 예외 형태가 출력될 수 있도록 초기 구성 (#5)

GlobalControllerAdvice.java
- 모든 패키지에서 공통으로 처리할 ExceptionHandler를 위해서만들었습니다.

SampleControllerAdive.java
- 샘플로 현재는 com.flab.marketgola.user 패키지와 그 하위 패키지에서 발생하는 예외를 처리하도록 구성하였습니다.

SampleErrorController.java, SampleErrorResult.java
- 실제로 예외메시지를 잘 보내는지 테스트하기 위해 추가하였습니다.

sgo8308 committed:
Merge branch 'feature/jiwoo_project_setting' into feature/jiwoo_errorpage_setting

derek-Flab commented on src/main/java/com/flab/marketgola/user/exception/SampleControllerAdvice.java:
```diff
@@ -0,0 +1,29 @@
+    @ResponseStatus(HttpStatus.BAD_REQUEST) // 상태코드 등록
+    @ExceptionHandler({NullPointerException.class}) // 이 핸들러로 처리하고자 하는 예외 등록
+    public SampleErrorResult exHandle(Exception e, HttpServletRequest request, ZoneId zoneId) {
+        log.error("예외 관련 메세지", e); //예외에 대한 로그
```
저희가 지금 API 서버를 만들고 있죠? 
예외가 발생했을 때 로그를 기록한다면, 어떤 요청이 들어왔을 때? 발생한 에러지? 정도는 로그를 보고 유추할 수 있도록 
요청 정보(요청 uri, http method) 와 같이 로그를 찍으면 좋을 것 같습니다.

derek-Flab commented on src/main/java/com/flab/marketgola/user/exception/SampleControllerAdvice.java:
```diff
@@ -0,0 +1,29 @@
+
+@Slf4j
+@RestControllerAdvice("com.flab.marketgola.user") // 예외 처리 범위 등록
+public class SampleControllerAdvice {
```
각 도메인 패키지 별로 Exception Handler를 등록하는 구조인거죠?
구현을 상상해보시죠. 
예를 들어 user를 조회했을 때 유저가 없다면 user가 없는 예외가 발생하겠죠?
UserNotFoundException을 정의해서 사용하면 되겠죠? 
이 각각의 exception의 종류는 패키지별로 정의해놓으면 좋을 것 같습니다. 

예외를 처리하는 handler부분은 common에서 공통적으로 처리하는게 좋지 않을까요? 

좀 더 구체적으로 요구사항을 정리해보면
1. 어떤 곳에서 발생하는 예외든지, 공통적으로 처리하는 핸들러가 있으면 좋겠다. 
2. FE에서는 약속된 포맷으로 발생된 예외 response를 받아 보면 좋겠다. 
```json
{
    "timestamp": "2022-06-19T14:45:29.723+00:00",
    "status": 400,
    "error": "Not Found",
    "errorCode": "UserNotFound",
    "message": "user not found"
}
```
3. 추가적으로 서버에서 서비스 관리를 위해 로깅을 했으면 좋겠는데 예외가 던져질 때 혹은 예외의 종류에 따라 로그 레벨을 달리하고 싶다. 예를 들면 UserNotFound는 debug레벨에서만 로그를 찍고 싶고 error 레벨에서는 찍고 싶지 않다. 또는 UserBannedException은 error 레벨로 찍고 싶다하는 요구사항이 만족이 되면 좋을 것 같습니다. (실제로 무시해도 되는 예외가 로그에 계속 찍히고, 심지어 로그 시스템에서 해당 로그 레벨이 찍히면 알람까지 오는 경우라면 공해나 다름없기 때문이죠)

접근방법추천) RuntimeException을 상속하는 BaseException? CommonException? 이라는 클래스를 정의하고 여기에 공통의 기능 및 필드를 정의하고, 이를 상속하여 구체화하는 예외 클래스를 만들면 좋을 것 같습니다. 그리고 이 예외 핸들러에서는 이 상위 예외 클래스를 잡아서 처리해주는 형태로요.

derek-Flab commented on src/main/java/com/flab/marketgola/user/exception/SampleControllerAdvice.java:
```diff
@@ -0,0 +1,29 @@
+        log.error("예외 관련 메세지", e); //예외에 대한 로그
+
+        String timestamp = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss")
+                .format(LocalDateTime.now(zoneId));
```
저희는 글로벌향 서비스가 아니라고 생각되어 타임존까지 고려한 처리는 무시해도 되지않을까 싶습니다. 한마디로 타임존은 고려하지 맙시다 라는 의견입니다.

derek-Flab replied:
추가적으로 지금 "yyyy-MM-dd HH:mm:ss"형태로 localdatetime -> string으로 변환하는 것이죠? 
이런 부분은 여기 뿐만 아니라 다른 곳에서도 충분히 중복해서 쓸 수 있을 것 같습니다. util성 기능이며 공통화 대상이라고 판단되요. 
`common` > `util` > `TimeUtil`클래스를 만들어서 유틸 클래스 안에서 처리하고 서비스 전반적으로 같이 쓰면 어떨까요

sgo8308 committed:
feat(user, common): BaseException과 FatalException 생성 (#5)

BaseException
* 추후 예외처리를 위해 BaseException 상속하는 모든 예외들은 자신과 관련된 응답 상태코드를 생성시 등록해야 함

FatalException
* Exception을 구분하기 위해 FatalException 생성
* Error level의 log를 남기고 싶은 에외는 FatalException을 상속하면 됨

sgo8308 committed:
feat(common): ErrorResult 변수 추가 (#5)

sgo8308 committed:
feat(common): TimeUtil 추가(#5)

LocalDataTime을 정돈된 형태의 String으로 변환하는 convert()추가

sgo8308 committed:
feat(user, common): 패키지 특화 ControllerAdvice 삭제 후 GlobalControllerAdvice 기능 추가 (#5)

exceptionHandler()
모든 예외를 BaseException을 상속한 예외로 감싸서 던질 시 모든 예외를 처리가능

doLog()
exception의 타입에 따라 다른 log level로 로깅

createErrorResult()
일관된 예외 결과를 만들어주며 예외가 갖고 있는 responseHttpStatus를 응답상태코드로 이용

sgo8308 committed:
feat(user): SampleErrorController mapping 추가, SampleErrorResult 삭제 (#5)

SampleErrorController
FatalException과 GeneralException을 테스트하기 위한 mapping 메소드 추가

SampleErrorResult
삭제 후 ErrorResult 추가

derek-Flab commented on src/main/java/com/flab/marketgola/common/util/TimeUtil.java:
```diff
@@ -0,0 +1,15 @@
+
+public class TimeUtil {
+
+    private TimeUtil() {
```
싱글톤 클래스를 만들어주는 것이죠? 
롬복의 `@UtilityClass`도 한번 찾아보실래요?

sgo8308 replied:
Utility클래스에 boilerplate 코드들을 작성해주는 어노테이션인 것을 확인했습니다.
그런데 현재 [롬복 공식 문서](https://projectlombok.org/features/experimental/UtilityClass#:~:text=Current%20status%3A%20negative%20%2D%20Currently%20we%20feel%20this%20feature%20cannot%20move%20out%20of%20experimental%20status%20due%20to%20fundamental%20issues%20with%20non%2Dstar%20static%20imports.)를 보면 이 기능은 non-star static imports에 대한 이슈가 있어서 실험 버전에서 정식 버전으로는 안 올라올 것 같다고 하는데 그래도 사용하는 게 좋을까요?

forest-lim-dev replied:
으흠 저같은 경우는 이전부터 사용하고 있긴 했는데, 레퍼런스 내용을 보니 안쓰는게 맞겠네요.!

derek-Flab replied:
저희 팀 같은 경우는 사용하고 있긴 한데요. 
레퍼런스 내용 검토로 사용하지 않는 것이 낫겠네요.

derek-Flab commented on src/main/java/com/flab/marketgola/common/exception/GlobalControllerAdvice.java:
```diff
@@ -0,0 +1,53 @@
+    /**
+     * 예외 처리 핸들러 BaseException 하위의 예외를 받아서 예외 결과를 일관되게 반환한다.
+     */
+    @ExceptionHandler()
```
BaseException 타입 일 때? 처리하는 핸들러로 지정하는게 좋겠어요!

만약 그 외(BaseException이 아닌 Exception) 인 경우 어떻게 처리하면 좋을까요?

sgo8308 replied:
그외 Exception도 어차피 BaseException을 상속한 직접 정의한 예외로 감싸면 되지 않을까?라고 생각했습니다. 그렇게 되면 BaseException을 처리하는 핸들러만으로도 모든 예외를 다 처리할 수 있다라고 생각했습니다.
결국 FE로 에러 결과를 반환할 때 NullPointerException이나 IllegarArgumentException같은 직접적인 예외보다는 좀 더 사용자 친화적인 예외로 감싸서 반환해야 하지 않나요?

그게 아니라면 모든 예외에 대해 일일이 HttpStatus나 logging level을 정해서 GlobalControllerAdvice에 각각의 예외를 처리하는 ExceptionHandler를 구현해야하지 않을까?라는 생각이 드는데 어떻게 하면 좋을까요?

derek-Flab replied:
@sgo8308 
NullPointerException, IllegarArgumentException 등등의 Exception은 저희가 한번 감싸서 사용자 지정 예외로 만들 수 있을 것 같습니다. 하지만, 라이브러리 등에서 발생하는 모든 예외를 저희가 커버할 수는 없을 것 이고, 저희가 BaseException으로 상속하지 않은 Exception은 따로 잡아서 internal server error로 처리해주면 되지 않을까요? 저희가 커버하지 못한 에러들이고, 이는 말그대로 서버 에러 라고 볼 수 있을 것 같습니다.

derek-Flab commented on src/main/java/com/flab/marketgola/common/exception/GlobalControllerAdvice.java:
```diff
@@ -0,0 +1,53 @@
+     * 예외 결과를 만들어주는 메소드
+     */
+    private ResponseEntity<ErrorResult> createErrorResult(BaseException e) {
+        String timestamp = TimeUtil.convert(LocalDateTime.now());
```
Time을 convert하는 아이디어는 이해했습니다. 그냥 localdatetime을 내리지 않고 변환하는 것은 어떤 의도일까요?

sgo8308 replied:
음.. 2022-06-19T14:45:29.723+00:00 이런 식으로 나오는 게 좀 보기 지저분해 보인다고 생각했었는데 딱히 그런 문제가 없다면 굳이 변환할 필요가 없을 것 같습니다 !

derek-Flab replied:
FE에서 localdatetime으로 내렸을 때 일관되게 변환하는 방법으로 코드를 구성할 수 있을 겁니다! 
localdatetime을 그대로 내리는 것으로도 충분할 것 같습니다.

derek-Flab commented on src/main/java/com/flab/marketgola/common/exception/FatalException.java:
```diff
@@ -0,0 +1,15 @@
+
+import org.springframework.http.HttpStatus;
+
+public class FatalException extends BaseException {
```
BaseException > FatalException > 구체적인 Exception을 선언하는 아이디어 군요!? 
약간 복잡한 듯 할수도 있지만 아이디어는 이해했습니다 굳! 
다만, FatalException을 상속한 UserNotFoundException이 있다고 예를 들어보죠. 
이 유저에게 배송을 하려고하는데 유저 정보가 없어서(예를들어 탈퇴 해버림) 발생한 예외는 fatal exception이 맞는 것 같습니다. 
근데, 유저가 단순히 댓글을 적고 탈퇴한 후 댓글을 쓴 유저를 접근하려고 할 때 발생하는 경우는 fatal은 아닐 수 있을 것 같습니다. 

이런 경우 같은 예외를 로직에 따라 던질때 로그레벨을 정하고 싶거나, 그런 니즈가 많을 것 같은데요~
이는 결국 생성할때 로그레벨을 지정할 수 있게 하면 좋겠다는 말이죠~

sgo8308 committed:
refactor(common, user): BaseException 수정, FatalException 삭제 (#9)

* LogLevel과 HttpStatus를 예외를 생성할 때 유연하게 구성할 수 있도록 변수를 추가 후 생성자 변경
* 바뀐 BaseException에 따라 SampleErrorController, SampleFatalException,
 SampleGeneralException 수정
* 불필요해진 FatalException은 제거

sgo8308 committed:
refactor(common): BaseException의 logLevel 변수에 따라 log level을 다르게 찍도록 변경(#9)

sgo8308 committed:
refactor(common): ErrorResult가 LocalDateTime을 그대로 포함하도록 변경 후 그에 따른 수정 (#9)

sgo8308 committed:
refactor(common): BaseException을 처리하는 핸들러의 이름과 처리 타입을 명확하게 변경 (#9)

sgo8308 committed:
feat(common): UnExpectedException 추가, BaseException 생성자 추가 (#9)

UnExpectedException
* BaseException을 상속하여 다른 예외들과 마찬가지로 일관되게 처리될 수 있도록 추가

BaseException
* 예상하지 못한 예외를 원인 예외로 등록할 수 있도록 생성자 추가

sgo8308 committed:
feat(common): GlobalControllerAdvice에 unExpectedHandler() 추가 (#9)

GlobalControllerAdvice
* UnExpectedException은 이 핸들러에서 다른 예외들과 같이 일관되게 처리

SampleErrorController
*  UnExpectedExceptionHandler를 테스트할 수 있는 핸들러 메소드 추가

derek-Flab commented on src/main/java/com/flab/marketgola/common/exception/GlobalControllerAdvice.java:
```diff
@@ -0,0 +1,66 @@
+            log.error("요청 uri = {}, http method = {}", request.getRequestURI(),
+                request.getMethod(), e);
+        } else {
+            log.debug("요청 uri = {}, http method = {}", request.getRequestURI(),
```
저희 프로젝트에서 로그레벨을 상세하게 나눌 필요가 있나 ? 좀 오버스럽다고 생각하실 수도 있겠지만요. 
실무에서는 꽤 여러개로 나눠서 사용하거든요. 
보통 debug, info, warn, error 정도는 분류해서 사용하고 있습니다. 
이 4개 정도는 지원해주는 것이 좋지 않겠나 싶습니다. 

debug : 개발 환경에서 다 확인해보고 싶을 때 
info: 에러가 아니지만 정보 용도로 남기는 ? 결제가 빈번하지 않은 서비스들은 이런건 로그로 남기고 적재시켜보겠다 할 수도있구요. 
warn: 에러는 아니지만 잠재적인 위험 요인들을 다루는데, 보통 스프링에서 발생하는 워닝 로그들은 꼼꼼히 봐주어야 할 필요가 있습니다. 요건 커스텀 예외에서 지원해줄 필요가 있을까? 생각이 들긴하지만, 뭐 사용은 하니까 지원해주면 좋을 것 같습니다. 
error: 이미 지원해주시고 계시지만, 이 로그는 반드시 처리해야합니다. 커스텀 예외에서도 그런 용도로 활용하고 있습니다.

derek-Flab commented on src/main/java/com/flab/marketgola/common/exception/GlobalControllerAdvice.java:
```diff
@@ -0,0 +1,66 @@
+     */
+    private void doLog(BaseException e, HttpServletRequest request) {
+        if (e.getLogLevel() == LogLevel.ERROR) {
+            log.error("요청 uri = {}, http method = {}", request.getRequestURI(),
```
이 중복 되는 에러 로그 메시지를 상수로 빼고, String.format()등을 사용해서 처리해보면 좋을 것 같습니다.

sgo8308 committed:
refactor(common): 로그레벨 4개로 나누고 중복되는 로그 메세지 변수로 빼기 (#9)
